File: eeducation/src/main/kotlin/rs/eeducation/service/CourseService.kt
package rs.eeducation.service

import org.springframework.stereotype.Service
import rs.eeducation.dto.CreateCourseDTO
import rs.eeducation.dto.EditCourseDTO
import rs.eeducation.model.Course
import rs.eeducation.model.School
import rs.eeducation.model.Student
import rs.eeducation.model.Teacher
import rs.eeducation.repository.CourseRepository
import java.text.SimpleDateFormat
import javax.persistence.EntityNotFoundException

@Service
class CourseService(private val courseRepository: CourseRepository,
                    private val teacherService: TeacherService,
                    private val schoolService: SchoolService,
                    private val userService: UserService) {

    fun getTeacherCourses(teacherId: Long): List<Course> {
        val teacher = teacherService.findById(teacherId)
        return courseRepository.findByTeachersContaining(teacher)
    }

    fun findById(courseId: Long): Course {
        return courseRepository.findById(courseId).orElseThrow { EntityNotFoundException("Course does not exist") }
    }

    fun save(course: Course): Course {
        return courseRepository.save(course)
    }

    fun createCourse(createCourseDTO: CreateCourseDTO): Course {
        val teachers = HashSet<Teacher>()
        for (teacher in createCourseDTO.teachers) {
            teachers.add(teacherService.findByEmail(teacher))
        }
        var school: School? = null
        if (createCourseDTO.school != -1L) {
            school = schoolService.findById(createCourseDTO.school)
        }
        val sdf = SimpleDateFormat("MM/dd/yyyy")
        val course = Course(null, HashSet(), HashSet(), teachers, school, HashSet(), HashSet(), HashSet(),
                createCourseDTO.name, createCourseDTO.begins, createCourseDTO.ends, false, createCourseDTO.freeToJoin, HashSet(), createCourseDTO.description)
        val savedCourse = courseRepository.save(course)
        for (teacher in teachers) {
            (teacher.courses as MutableSet).add(savedCourse)
            teacherService.save(teacher)
        }
        return savedCourse
    }

    fun editCourse(editCourseDTO: EditCourseDTO): Course {
        //Get course
        val course = courseRepository.findById(editCourseDTO.id).orElseThrow { EntityNotFoundException("Course does not exist") }
        //get logged in user
        val user = userService.getLoggedInUser()
        //Check if editor is school or teacher of said course if not throw exception
        userService.teacherTeacherCourse(user!!, course)
        //edit course
        course.name = editCourseDTO.name
        course.beginsDate = editCourseDTO.begins
        course.endsDate = editCourseDTO.ends
        course.descr = editCourseDTO.description
        //save course
        //return course
        return courseRepository.save(course)
    }

    fun archiveCourse(courseId: Long): Course {
        val course = courseRepository.findById(courseId).orElseThrow { EntityNotFoundException("Course not exist") }
        val user = userService.getLoggedInUser()
        userService.teacherTeacherCourse(user!!, course)
        course.isArchivated = true
        return courseRepository.save(course)
    }

    fun joinCourse(courseId: Long): Course {
        val student = userService.getLoggedInUser() as Student
        val course = findById(courseId)
        if (course.freeToJoin) {
            (course.students as MutableSet).add(student)
        } else {
            (course.studentsWaiting as MutableSet).add(student)
        }
        return save(course)
    }

    fun findStudentsOfClass(courseId: Long): List<Student> {
        val course = findById(courseId)
        return course.students.toList()
    }
}

File: eeducation-front/src/app/service/course.service.ts
import {Injectable} from '@angular/core';
import {CreateCourse} from '../model/CreateCourse';
import {Observable} from 'rxjs';
import {Course} from '../model/Course';
import {HttpClient} from '@angular/common/http';
import {PathService} from './path.service';
import {CourseBasic} from '../model/CourseBasic';
import {EditCourse} from '../model/EditCourse';
import {Student} from '../model/Student';

@Injectable({
  providedIn: 'root'
})
export class CourseService {
  private path = '';

  constructor(private http: HttpClient, private pathService: PathService) {
    this.path = pathService.path + '/api/courses';
  }

  addCourse(createCourse: CreateCourse): Observable<Course> {
    return this.http.post<Course>(this.path, createCourse);
  }

  getTeacherCourses(): Observable<Array<CourseBasic>> {
    return this.http.get<Array<CourseBasic>>(this.path + '/teacher' );
  }

  getCourse(courseId: number): Observable<Course> {
    return this.http.get<Course>(this.path + '/' + courseId);
  }

  editCourse(editCourse: EditCourse): Observable<Course> {
    return this.http.put<Course>(this.path, editCourse);
  }

  archiveCourse(courseId: number): Observable<Course> {
    return this.http.delete<Course>(this.path + '/' + courseId);
  }

  joinCourse(courseId: number): Observable<Course> {
    return this.http.get<Course>(this.path + '/join/' + courseId);
  }

  viewStudentsOfClass(courseId: number): Observable<Array<Student>> {
    return this.http.get<Array<Student>>(this.path + '/students/' + courseId);
  }

}
